File: 230.二叉搜索树中第k小的元素.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int val;
    int cnt = 0;
    // 中序遍历
    int kthSmallest(TreeNode* root, int k) {
        inOrder(root, k);
        return val;
    }

    void inOrder(TreeNode* root, int k){
        if(!root)   return;
        inOrder(root->left, k);
        ++cnt;
        if(cnt == k){
            val = root->val;
            return;
        }
        inOrder(root->right, k);
    }
};

File: 树/230.二叉搜索树中第K小的元素.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {

    const getCount = (root) => {
        if (root == null)
            return 0;
        return 1 + getCount(root.left) + getCount(root.right);
    }

    while (true) {
        let l = getCount(root.left);
        if (l >= k) {
            root = root.left;
        } else if (l == k - 1)
            return root.val;
        else {
            root = root.right;
            k = k - l - 1;
        }
    }
};

var kthSmallest = function (root, k) {
    let count = 0;
    let res;

    const inorder = (root) => {
        if (root === null)
            return;
        inorder(root.left);
        ++count;
        if (count === k) {
            res = root.val;
            return;
        }
        inorder(root.right);
    }

    inorder(root)
    return res;
}